cmd/gaia/app/export.go

```diff
@@ -0,0 +1,156 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	distr "github.com/cosmos/cosmos-sdk/x/distribution"
+	"github.com/cosmos/cosmos-sdk/x/gov"
+	"github.com/cosmos/cosmos-sdk/x/mint"
+	"github.com/cosmos/cosmos-sdk/x/slashing"
+	stake "github.com/cosmos/cosmos-sdk/x/stake"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// export the state of gaia for a genesis file
+func (app *GaiaApp) ExportAppStateAndValidators(forZeroHeight bool) (
+	appState json.RawMessage, validators []tmtypes.GenesisValidator, err error) {
+
+	// as if they could withdraw from the start of the next block
+	ctx := app.NewContext(true, abci.Header{Height: app.LastBlockHeight()})
+
+	if forZeroHeight {
+		prepForZeroHeightGenesis(ctx)
+	}
+
+	// iterate to get the accounts
+	accounts := []GenesisAccount{}
+	appendAccount := func(acc auth.Account) (stop bool) {
+		account := NewGenesisAccountI(acc)
+		accounts = append(accounts, account)
+		return false
+	}
+	app.accountKeeper.IterateAccounts(ctx, appendAccount)
+
+	genState := NewGenesisState(
+		accounts,
+		auth.ExportGenesis(ctx, app.feeCollectionKeeper),
+		stake.ExportGenesis(ctx, app.stakeKeeper),
+		mint.ExportGenesis(ctx, app.mintKeeper),
+		distr.ExportGenesis(ctx, app.distrKeeper),
+		gov.ExportGenesis(ctx, app.govKeeper),
+		slashing.ExportGenesis(ctx, app.slashingKeeper),
+	)
+	appState, err = codec.MarshalJSONIndent(app.cdc, genState)
+	if err != nil {
+		return nil, nil, err
+	}
+	validators = stake.WriteValidators(ctx, app.stakeKeeper)
+	return appState, validators, nil
+}
+
+// prepare for fresh start at zero height
+func prepForZeroHeightGenesis(ctx sdk.Context) {
+
+	/* TODO XXX check some invariants */
```

Comment:
can we not just reuse the simulation invariants? We should either resolve this or open an issue instead of using `XXX` I try to remove all the `XXX` before each merge

Reply:
I agree but we need a larger refactor of the simulation invariants to reuse them here (since they would need to take in a particular `sdk.Context` instead of a `BaseApp`, which they do at the moment) - spawned a new issue: https://github.com/cosmos/cosmos-sdk/issues/2885.

Reply:
We can remove this check altogether when we fix the underyling issue, it's not necessary since all the simulation invariants are checked when state is *imported*.

Reply:
:+1: thanks for opening the issue